shopinvader_api_sale_loyalty/routers/loyalty.py

```diff
@@ -0,0 +1,186 @@
+
+    # Get rewards
+    @api.model
+    def _get_rewards(self, code: str):
+        card = self.env["loyalty.card"].search([("code", "=", code)], limit=1)
+        program_id = self.env["loyalty.program"]
+        if card:
+            program_id = card.program_id
+        else:
+            rule_id = self.env["loyalty.rule"].search([("code", "=", code)], limit=1)
+            if rule_id:
+                program_id = rule_id.program_id
+        if program_id:
+            return program_id.reward_ids
+        return self.env["loyalty.reward"]
+
+    # Apply coupon
+    @api.model
+    def _check_code_and_reward(self, cart, code, reward_id=None):
+        """
+        * Check the code is valid.
+        * Check the provided reward is allowed.
+          If no reward is given, return the unique allowed reward,
+          or raise an error.
+
+          :return: the browsed reward
+        """
+        status = cart._try_apply_code(code)
+        if "error" in status:
+            raise UserError(status["error"])
+        all_rewards = self.env["loyalty.reward"]
+        for rewards in status.values():
+            all_rewards |= rewards
+
+        if not all_rewards:
+            # TODO: do we want to raise a UserError, or a specific error with error code?
+            raise UserError(_("No reward available for this code."))
+        if reward_id and reward_id not in all_rewards.ids:
+            raise UserError(_("Reward not allowed for this code."))
+        if not reward_id and len(all_rewards) > 1:
+            raise UserError(_("Several rewards available. Please specify one."))
+
+        if reward_id:
+            return self.env["loyalty.reward"].browse(reward_id)
+        return all_rewards[0]  # In this case only 1 reward available
+
+    @api.model
+    def _check_free_reward_product(self, reward, product_id=None):
+        """
+        Check the free product is allowed (if given) or check
+        that it is unique (if not given).
+
+        :return: the free product id
+        """
+        if not reward.reward_type == "product":
+            return
+
+        reward_products = reward.reward_product_ids
+
+        if not reward_products:
+            raise UserError(_("No free products available."))
+        if product_id and product_id not in reward_products.ids:
+            raise UserError(_("Free product not allowed for this reward."))
+        if not product_id and len(reward_products) > 1:
+            raise UserError(_("Several free products available. Please specify one."))
+
+        if product_id:
+            return product_id
+        return reward_products[0].id
+
+    @api.model
+    def _apply_coupon(self, cart: "SaleOrder", data: LoyaltyCardInput):
+        """Apply a coupon or promotion code.
+
+        It can raise UserError if coupon is not applicable, or
+        if the coupon let the choice between several rewards, and the
+        selected reward is not specified.
+        """
```

Comment:
In v16 a given coupon code can let the customer chose between several rewards.
If calling the `apply_coupon` service on such a loyalty program, without specifying which reward to apply, the request cannot succeed.
The methods in `shopinvader_api_loyalty.service.helper` raise a `UserError` in such cases. Should the service catch these errors to return a cleaner error with a specific error code?